openwisp_controller/connection/tests/test_models.py

```diff
@@ -739,30 +739,79 @@ def test_device_config_created(self, mocked_sleep, mocked_connect):
 
+        with self.subTest('openwisp_config >= 0.6.0a'):
+            conf.config = '{"dns_servers": []}'
+            conf.full_clean()
+            with mock.patch(_exec_command_path) as mocked_exec_command:
+                mocked_exec_command.return_value = self._exec_command_return_value(
+                    stdout='openwisp_config 0.6.0a'
+                )
+                conf.save()
+                self.assertEqual(mocked_exec_command.call_count, 2)
+                _assert_version_check_command(mocked_exec_command)
+                args, _ = mocked_exec_command.call_args_list[1]
+                self.assertIn('OW_CONFIG_PID', args[0])
             conf.refresh_from_db()
             self.assertEqual(conf.status, 'applied')
 
-        with self.subTest('exit_code 1'):
+        with self.subTest('openwisp_config < 0.6.0a: exit_code 0'):
             conf.config = '{"interfaces": []}'
             conf.full_clean()
             with mock.patch(_exec_command_path) as mocked_exec_command:
                 mocked_exec_command.return_value = self._exec_command_return_value(
-                    exit_code=1
+                    stdout='openwisp_config 0.5.0'
                 )
                 conf.save()
                 self.assertEqual(mocked_exec_command.call_count, 2)
+                _assert_version_check_command(mocked_exec_command)
+                _assert_applying_conf_test_command(mocked_exec_command)
+            conf.refresh_from_db()
+            self.assertEqual(conf.status, 'applied')
 
+        with self.subTest('openwisp_config < 0.6.0a: exit_code 1'):
+            conf.config = '{"radios": []}'
+            conf.full_clean()
+            with mock.patch(_exec_command_path) as mocked_exec_command:
+                stdin, stdout, stderr = self._exec_command_return_value(
+                    stdout='openwisp_config 0.5.0'
+                )
+                # An iterable side effect is required for different return exit codes:
+                # 1. Checking openwisp_config returns with 0
+                # 2. Testing presence of /tmp/openwisp/applying_conf returns with 1
+                # 3. Restarting openwisp_config returns with 0 exit code
+                stdout.channel.recv_exit_status.side_effect = [0, 1, 0]
+                mocked_exec_command.return_value = (stdin, stdout, stderr)
+                conf.save()
+                self.assertEqual(mocked_exec_command.call_count, 3)
+                _assert_version_check_command(mocked_exec_command)
+                _assert_applying_conf_test_command(mocked_exec_command)
+                args, _ = mocked_exec_command.call_args_list[2]
+                self.assertEqual(args[0], '/etc/init.d/openwisp_config restart')
             conf.refresh_from_db()
-            # exit code 1 considers the update not successful
-            self.assertEqual(conf.status, 'modified')
+            self.assertEqual(conf.status, 'applied')
```

Comment:
this change is suspicious, why is this happening? Shouldn't keep working as before?

Reply:
I reverted it back. I wanted to make all four scenarios consistent. I changed the exit code in mock.